scylla/src/transport/load_balancing/default.rs

```diff
@@ -1563,4 +1583,315 @@ mod tests {
+            },
+            Test {
+                // Keyspace with RF=3 with disabled DC failover
+                policy: DefaultPolicy {
+                    preferred_datacenter: Some("eu".to_owned()),
+                    is_token_aware: true,
+                    permit_dc_failover: false,
+                    ..Default::default()
+                },
+                query_info: QueryInfo {
+                    token: Some(Token { value: 160 }),
+                    keyspace: Some(KEYSPACE_NTS_RF_3),
+                    consistency: Consistency::Quorum,
+                    ..Default::default()
+                },
+                // going though the ring, we get order: F , A , C , D , G , B , E
+                //                                      us  eu  eu  us  eu  eu  us
+                //                                      r2  r1  r1  r1  r2  r1  r1
+                expected_groups: ExpectedGroupsBuilder::new()
+                    .group([A, C, G]) // pick + fallback local replicas
+                    .group([B]) // local nodes
+                    .build(), // failover explicitly forbidden
+            },
+            Test {
+                // No token implies no token awareness
+                policy: DefaultPolicy {
+                    preferred_datacenter: Some("eu".to_owned()),
+                    is_token_aware: true,
+                    permit_dc_failover: true,
+                    ..Default::default()
+                },
+                query_info: QueryInfo {
+                    token: None, // no token
+                    keyspace: Some(KEYSPACE_NTS_RF_3),
+                    consistency: Consistency::Quorum,
+                    ..Default::default()
+                },
+                expected_groups: ExpectedGroupsBuilder::new()
+                    .group([A, B, C, G]) // local nodes
+                    .group([D, E, F]) // remote nodes
+                    .build(),
+            },
+            Test {
+                // No keyspace implies no token awareness
+                policy: DefaultPolicy {
+                    preferred_datacenter: Some("eu".to_owned()),
+                    is_token_aware: true,
+                    permit_dc_failover: true,
+                    ..Default::default()
+                },
+                query_info: QueryInfo {
+                    token: Some(Token { value: 160 }),
+                    keyspace: None, // no keyspace
+                    consistency: Consistency::Quorum,
+                    ..Default::default()
+                },
+                expected_groups: ExpectedGroupsBuilder::new()
+                    .group([A, B, C, G]) // local nodes
+                    .group([D, E, F]) // remote nodes
+                    .build(),
+            },
+            Test {
+                // Unknown preferred DC, failover permitted
+                policy: DefaultPolicy {
+                    preferred_datacenter: Some("au".to_owned()),
```

Comment:
Could this be set to `None`?

Reply:
Of course, but the very goal of this test is to check the DP's behaviour when `preferred_datacenter` is set to an unknown DC.

Reply:
I think it would be good to add a test where `preferred_datacenter` is None, I don't see one.